### ISSUE TITLE
Determine vector where canvasgeometries intersect

### ISSUE THREAD

**Issue by Karlitos2312** [Microsoft/Win2D#232]
Hi,
I have created 2 canvasgeometries - one is a line and the other is a rectangle. How can I determine the vector where the line intersects with the rectangle.

**Comment by clandrew** [comment]
Win2D doesn't have a feature for obtaining vectors-at-points-of-intersection, but it does have [Geometry.CombineWith](http://microsoft.github.io/Win2D/html/M_Microsoft_Graphics_Canvas_Geometry_CanvasGeometry_CombineWith_1.htm) using CanvasGeometryCombine.Intersect. This will yield a geometry that's the intersection of the two source geometries.

You can take that result geometry, and look at its points using [SendPathTo](http://microsoft.github.io/Win2D/html/M_Microsoft_Graphics_Canvas_Geometry_CanvasGeometry_SendPathTo.htm). Depending on the kind of information you need, you can compute the outermost points of intersection, or do something else with the line segment. 

Does this help?

**Comment by Karlitos2312** [comment]
Thanks for the swift response - can you give an idea of how to implement the SendPathTo method? I don't where ICanvasPathReceiver streamReader parameter comes from.

**Comment by clandrew** [comment]
Sure. Your app needs to define a class that implements ICanvasPathReceiver, construct an object of that type and pass it to SendPathTo. 

For a simple example in C#, see this file:
[https://github.com/Microsoft/Win2D/blob/master/winrt/test.managed/Shared/PathReceiver.cs](https://github.com/Microsoft/Win2D/blob/master/winrt/test.managed/Shared/PathReceiver.cs)

**Comment by Karlitos2312** [comment]
Thanks again - unfortunately I'm obviously missing something, as I can't see how the c# example would work for my purposes. As far as I can tell in the sample - the verify function would confirm an arc was created ( but  not returning any previous unknowns).

To expand on my app a bit further to perhaps make my issue clearer:

I have a polygon shape on screen and a circle point (pt1) elsewhere on screen - I have a line from the pt1 (line1) direct to the centre of the polygon. I have created a rectangle (rectBounds) boundary surrounding the polygon. I want to determine the intersection point from line1 with rectbounds.

combinedGeometry = RectGeometry.CombineWith(oLine, CanvasGeometryCombine.Intersect)

I tried computing the bounds of the combinedgeometry to work out the intersection point - but the resulting Rect returns infinite as it's values.

Apologies if I'm slow on the uptake here...

**Comment by clandrew** [comment]
This code example was simply to show how to define a class which implements ICanvasPathReceiver. Your app, as it hooks BeginFigure/AddLine, can do something with the points that it sees.

Drawing combinedGeometry might be able to tell you whether that intersection is what you'd expect.

**Comment by Karlitos2312** [comment]
I have drawn the combinedgeometry and the intersection is correct, my stumbling block is determining the length of the combinedgeometry.

**Comment by clandrew** [comment]
Ok. As for why combinedGeometry is infinite- I don't have enough information to determine that here. If you'd like more help with that part, please feel free to share out a fuller repro.

Or if SendPathTo doesn't help, let me know.

**Comment by Karlitos2312** [comment]
CanvasGeometry Poly = CanvasGeometry.CreatePolygon(sender.Device, pts);
Rect OrigShapeBounds = Poly.ComputeBounds();
CanvasGeometry RectGeometry = CanvasGeometry.CreateRectangle(sender.Device, OrigShapeBounds);

Matrix3x2 oRotationMatrix3x2 = Matrix3x2.CreateRotation((Single)nRotationRadChange, ShapeScreenGPS);
CanvasGeometry RotatedGeometry = Poly.Transform(oRotationMatrix3x2);

Rect rect2 = RotatedGeometry.ComputeBounds();
RectGeometry = CanvasGeometry.CreateRectangle(sender.Device, rect2);

oRotationMatrix3x2 = Matrix3x2.CreateRotation((Single)(nRotationRadChange * -1), ShapeScreenGPS);
CanvasGeometry RectGeometry2 = RectGeometry.Transform(oRotationMatrix3x2);
Rect RectBounds = RectGeometry2.ComputeBounds();

Vector2 ShapeCentreVector = new Vector2((Single)(RectBounds.X + RectBounds.Width / 2), (Single)(RectBounds.Y + RectBounds.Height / 2));
CanvasGeometry ShapeCentreCircle = CanvasGeometry.CreateCircle(sender.Device, ShapeCentreVector, 3);

double teeToCentreShapeMtrs = App.HoleMapViewModel.GetGPSdistance(TeeGPS, ShapeCentreVector) * 1000;
Single teeToCentreShapePixels = App.HoleMapViewModel.calculateDistance(CurrentScreenGPS, ShapeCentreVector);
teeToCentreShapePixels += (Single)(OrigShapeBounds.Height / 2);

CanvasPathBuilder ShapePathbuilder = new CanvasPathBuilder(sender.Device);
ShapePathbuilder.BeginFigure(CurrentScreenGPS);
ShapePathbuilder.AddLine(ShapeCentreVector);
ShapePathbuilder.EndFigure(CanvasFigureLoop.Open);

TeeToShapeLine = CanvasGeometry.CreatePath(ShapePathbuilder);
//clds.DrawGeometry(TeeToShapeLine, Colors.Cyan);


Matrix3x2 interGeometryTransform = Matrix3x2.CreateTranslation(new Vector2(0f));
CanvasGeometry combinedGeometry = RectGeometry2.CombineWith(TeeToShapeLine, interGeometryTransform, CanvasGeometryCombine.Intersect);

//double shapeHalfLength = CombinedGeometryBounds.Width;
clds.DrawGeometry(combinedGeometry, Colors.Red, 5);

Rect CombinedGeometryBounds = combinedGeometry.ComputeBounds();

**Comment by clandrew** [comment]
Some of the data here is missing.
- What is 'pts'?
- What is nRotationRadChange, ShapeScreenGPS and CurrentScreenGPS?

**Comment by Karlitos2312** [comment]
pts is an array of vector points (creates a polygon - any polygon will do)

  RotationRadians = Math.Atan2(CurrentScreenGPS.Y - ShapeScreenGPS.Y, CurrentScreenGPS.X - ShapeScreenGPS.X);
                        double nRotationRadChange = ConvertToRadians(90) - RotationRadians;

CurrentScreenGps is a point on the screen  - any point on the screen not important where it is.
ShapeScreenGPS is the Centre point of the polygon shape.

RotationRadians is the Radians angle between CurrentScreenGPS and ShapeScreenGPS

**Comment by 13thsymphony** [comment]
It might be easier to share a minimal repro project with us?

**Comment by clandrew** [comment]
Could you please provide actual data values for these? Or, a standalone project.

**Comment by Karlitos2312** [comment]
Are the files attached sufficient?

[ShapesTest.zip](https://github.com/Microsoft/Win2D/files/83359/ShapesTest.zip)

**Comment by clandrew** [comment]
These files just declare a polygon- they don't have anything to do with rectangles or intersecting them? (Unless I'm missing something)

Could you please attach a project which we can build and run, and performs your geometry intersect?

**Comment by Karlitos2312** [comment]
[Win2dShapeTest.zip](https://github.com/Microsoft/Win2D/files/83381/Win2dShapeTest.zip)

**Comment by clandrew** [comment]
Great, thanks for sending this along. The repro worked for me.

If we look carefully at the bounds, they are not actually infinite. They are:
X=Inf, Inf, Width=-Inf, Height=-Inf

It's a little unintuitive but these are actually Direct2D's 'canonically empty' bounds, meaning that the intersection has zero size. 

Why does this intersection have zero bounds? Because ComputeBounds only measures *fill*. In this app, you're doing an intersection between a rectangle and a line, and that has no fill.

ComputeBounds is not the way to go here.

Instead, you should send your geometry to a path receiver and look at the points it streams back. In this particular example, define this class:

`
        private class MyGeometryStreamReader : ICanvasPathReceiver
        {
            public List<Vector2> Points;

            public MyGeometryStreamReader()
            {
                Points = new List<Vector2>();
            }

            public void BeginFigure(Vector2 point, CanvasFigureFill fill)
            {
                Points.Add(point);
            }
            public void AddArc(Vector2 point, float x, float y, float z, CanvasSweepDirection sweepDirection, CanvasArcSize arcSize)
            {
            }
            public void AddCubicBezier(Vector2 controlPoint1, Vector2 controlPoint2, Vector2 endPoint)
            {
            }
            public void AddLine(Vector2 endPoint)
            {
                Points.Add(endPoint);
            }
            public void AddQuadraticBezier(Vector2 controlPoint, Vector2 endPoint)
            {
            }

            public void SetFilledRegionDetermination(CanvasFilledRegionDetermination filledRegionDetermination)
            {
            }

            public void SetSegmentOptions(CanvasFigureSegmentOptions figureSegmentOptions)
            {
            }

            public void EndFigure(CanvasFigureLoop figureLoop)
            {
            }
        }
`

And then in your code, below the intersection test, you would use:

                var streamReader = new MyGeometryStreamReader();
                combinedGeometry.SendPathTo(streamReader);
                Vector2 point1 = streamReader.Points[0];
                Vector2 point2 = streamReader.Points[1];
                double cyanLineLength = Math.Sqrt(Math.Pow(point1.X - point2.X, 2) + Math.Pow(point1.Y - point2.Y, 2));
`

**Comment by Karlitos2312** [comment]
Hi,
Thanks very much for this - I would never have figured this out on my own. 
For my future development - can you talk me through a few things on the MyGeometryStreamReader class:

1. Are the BeginFigure and AddLine methods used because its building a path?
2. Where are the BeginFigure and AddLine methods actually called?
3. Can there be more than 2 points (start & end point) returned from streamreader?

Please be aware - This level of support from you has made my day!!! - you should be very proud, keep it up....

(You may have just opened Pandora's Box)

**Comment by clandrew** [comment]
Sure. 

MyGeometryStreamReader is for Win2D to tell your app about the data for a path. It doesn't build a path, it just tells you about the contents of an already-existing path.

1. Since your path contains one figure and a line, BeginFigure gets called once, and AddLine gets called once. EndFigure gets called, too, but we've ignored it here.

2. Win2D calls them. It calls them when your app calls SendPathTo. 

3. Yes, depending on the geometry we've passed to SendPathTo. In the case of your specific app, it is pretty safe to assume there are two points, or no points if it doesn't intersect at all, since your line geometry is just two points. This assumption may no longer hold, if you start working with more complicated figures.

I'm really glad this helped. This seems to have resolved everything, so I'll close the issue.
